Custom Entities Guide

JaaiDead opened the pull request:
closes #147

https://deploy-preview-267--nimble-elf-d9d491.netlify.app/develop/entities/first-entity

---

TODO [by Earthcomputer](https://github.com/FabricMC/fabric-docs/pull/267#issuecomment-2691246954)

- [x] Create and register entity and attributes.
- [x] Add and explain goals
- [x] And renderer, model and texture
   - This will include the leg swing animation, but not the dancing animation.
   - At this point, the mob can be spawned with  `/summon`.
- [x] Add `TrackedData` for dancing, and explain how to use the data tracker for synchronizing values to the client.
- [x] Add `dancingTimeLeft` which will be saved and loaded from NBT, explain saving and loading entities from NBT.
- [x] Add the dancing animation.
- [x] Add the spawn egg.

JaaiDead committed:
Disable lastUpdated in VitePress config and add custom entities documentation and classes

JaaiDead committed:
Implement custom entity in server-side.

JaaiDead commented:
Implementation of Client-Side Code needs suggestion , Server Side Code - needs review

JaaiDead committed:
Merge branch 'main' into entities

JaaiDead committed:
Merge branch 'main' into entities

JaaiDead committed:
Merge branch 'main' into entities

JaaiDead committed:
Merge branch 'main' into entities

JaaiDead committed:
Merge branch 'FabricMC:main' into entities

JaaiDead committed:
Merge branch 'main' into entities

JaaiDead committed:
Cleanup and TODO

JaaiDead committed:
Merge branch 'entities' of https://github.com/JaaiDead/fabric-docs into entities

JaaiDead committed:
Merge branch 'FabricMC:main' into entities

Earthcomputer committed:
Implement mini-golem entity

Earthcomputer committed:
Add myself to custom-entities.md credits

JaaiDead committed:
Merge pull request #2 from Earthcomputer/mini-golem

Add tutorial entity code

JaaiDead committed:
Merge branch 'FabricMC:main' into entities

JaaiDead committed:
Add Sidebar Translations

its-miroma commented on develop/entities/custom-entities.md:
```diff
@@ -0,0 +1,9 @@
+  - Earthcomputer
+---
+
+## Custom Entities {#custom-entities}
```
```suggestion
```

no need ;)

JaaiDead committed:
Intro and TODO

JaaiDead commented:
Kindly check the to do on top on the recent commit and if im missing anything in the flow or the order is wrong , suggest changes.

Earthcomputer commented:
In my mind, the flow of the tutorial would be more like:

- [ ] Create and register entity and attributes.
   - This is the minimum amount of work to be able to run the `/summon` command, I believe.
- [ ] Add and explain goals
- [ ] And renderer, model and texture
   - This will include the leg swing animation, but not the dancing animation.
- [ ] Add `TrackedData` for dancing, and explain how to use the data tracker for synchronizing values to the client.
- [ ] Add `dancingTimeLeft` which will be saved and loaded from NBT, explain saving and loading entities from NBT.
- [ ] Add the dancing animation.
- [ ] Add the spawn egg.

its-miroma commented on develop/entities/custom-entities.md:
```diff
@@ -0,0 +1,23 @@
+
+The MiniGolemEntity is a custom mob based on `PathAwareEntity`.
+
+@[code lang=java transcludeWith=:::1](@reference\latest\src\main\java\com\example\docs\entity\MiniGolemEntity.java)
```
```suggestion
@[code lang=java transcludeWith=:::1](@/reference/latest/src/main/java/com/example/docs/entity/MiniGolemEntity.java)
```

JaaiDead committed:
Update develop/entities/custom-entities.md

Co-authored-by: Miroma <136986257+its-miroma@users.noreply.github.com>

JaaiDead committed:
Structure

JaaiDead committed:
Merge branch 'main' into entities

its-miroma commented on reference/latest/src/main/java/com/example/docs/entity/MiniGolemEntity.java:
```diff
@@ -0,0 +1,106 @@
+import net.minecraft.nbt.NbtCompound;
+import net.minecraft.recipe.Ingredient;
+import net.minecraft.world.World;
+//:::1
```
just a reminder early on: please use descriptive labels for transclusions, see #247

JaaiDead committed:
transclude nameing changes

JaaiDead committed:
Merge branch 'entities' of https://github.com/JaaiDead/fabric-docs into entities

JaaiDead committed:
Rendering

JaaiDead committed:
Rendering Setup

JaaiDead committed:
Merge branch 'main' into entities

JaaiDead commented:
Needs Code Rework for the latest version.

Which might change the flow.

cassiancc committed:
Merge branch 'main' into pr/267

cassiancc committed:
Port to 1.21.10 and Mojang Mappings

cassiancc committed:
checkstyle

its-miroma commented on reference/1.21.4/src/main/generated/assets/fabric-docs-reference/items/mini_golem_spawn_egg.json:
Why are we changing older versions?

its-miroma commented on develop/entities/custom-entities.md:
```diff
@@ -0,0 +1,83 @@
+---
+title: Custom Entities
+description: Learn how to do custom entities.
+authors:
+  - JaaiDead
+  - Earthcomputer
+---
```
```suggestion
authors:
  - cassiancc
  - Earthcomputer
  - JaaiDead
---

```

its-miroma commented on .vitepress/sidebars/develop.ts:
```diff
@@ -144,6 +144,10 @@ export default [
         link: "/develop/entities/damage-types",
       },
+      {
+        text: "develop.entities.custom-entities",
```
```suggestion
        text: "develop.entities.custom_entities",
```

its-miroma commented on sidebar_translations.json:
```diff
@@ -37,6 +37,8 @@
   "develop.blocks.block_entity_renderer": "Block Entity Renderers",
   "develop.entities": "Entities",
+  "develop.entities.damage-types": "Damage Types",
+  "develop.entities.custom-entities": "Custom Entities",
```
```suggestion
  "develop.entities.custom_entities": "Custom Entities",
```

**Damage Types** is below

its-miroma commented on develop/entities/custom-entities.md:
```diff
@@ -0,0 +1,83 @@
+---
+title: Custom Entities
```
Should this be **Creating Your First Entity**? and the filename

its-miroma reviewed (changes requested):
the article still needs work

JaaiDead committed:
Update develop/entities/custom-entities.md

Co-authored-by: Miroma <its.miroma@proton.me>

JaaiDead committed:
Update .vitepress/sidebars/develop.ts

Co-authored-by: Miroma <its.miroma@proton.me>

JaaiDead committed:
Update sidebar_translations.json

Co-authored-by: Miroma <its.miroma@proton.me>

JaaiDead committed:
Merge branch 'FabricMC:main' into entities

cassiancc committed:
remap article and review

JaaiDead commented:
can PR #320 be merged with this PR?

JaaiDead committed:
Create Basic Structure

cassiancc committed:
Merge branch 'main' into pr/267

cassiancc committed:
Avoid changing old versions

cassiancc committed:
re-fix old versions

cassiancc committed:
structural push

cassiancc committed:
Update first-entity.md

cassiancc committed:
fix checkstyle

cassiancc committed:
Fix wording on renderer

JaaiDead committed:
Merge branch 'main' into entities

JaaiDead committed:
Textures update for eggs

cassiancc committed:
Additional images

JaaiDead committed:
Change Spelling

JaaiDead committed:
Cleanup

Earthcomputer commented on develop/entities/first-entity.md:
```diff
@@ -0,0 +1,163 @@
+
+## Your First Entity {#your-first-entity}
+
+This tutorial will walk you through the process of creating a custom _Mini Iron Golem_. This entity will have fun animations.
```
```suggestion
This tutorial will walk you through the process of creating a custom _Mini Golem_. This entity will have fun animations.
```

Earthcomputer commented on develop/entities/first-entity.md:
```diff
@@ -0,0 +1,163 @@
+
+## Adding Goals {#adding-goals}
+
+Goals are the system that handle an entity's objective/aim, providing them with a defined set of behaviour. Goals have a certain priority, ranging from `0` being the highest and subsequently reducing in priority as the number increases.
```
```suggestion
Goals are the system that handle an entity's objective/aim, providing them with a defined set of behaviour. Goals have a certain priority: goals with a lower value for the priority are prioritized over goals with a higher value for the priority.
```

Earthcomputer commented on develop/entities/first-entity.md:
```diff
@@ -0,0 +1,163 @@
+To register your entity, it's recommended to create a separate class, `ModEntityTypes`, where you register any and all entity types, set their dimensions, and register their attributes.
+
+@[code transcludeWith=:::types](@/reference/latest/src/main/java/com/example/docs/entity/ModEntityTypes.java)
+
```
At this point you can already create your entity with `/summon` (although you won't be able to see it). I would maybe include this here as people like to see stuff working for their effort.

cassiancc replied:
As I understand it, creating the entity at this point causes either 1. a game crash when it loads since there is no renderer or 2. the text "failed to summon entity"

Earthcomputer replied:
Ah okay, in that case ignore me

Earthcomputer commented on develop/entities/first-entity.md:
```diff
@@ -0,0 +1,163 @@
+
+### Setting up the Model {#setting-up-model}
+
+The `MiniGolemEntityModel` class defines how your entity looks by describing its shape and parts. Models are generally created in third-party tools like [Blockbench](https://web.blockbench.net), rather than being written by hand.
```
```suggestion
The `MiniGolemEntityModel` class defines how your entity looks by describing its shape and parts. Models are generally created in third-party tools like [Blockbench](https://web.blockbench.net), rather than being written by hand. Nonetheless, this tutorial will go through a manual example to show you how it works.
```

Earthcomputer commented on develop/entities/first-entity.md:
```diff
@@ -0,0 +1,163 @@
+
+@[code transcludeWith=:::model1](@/reference/latest/src/client/java/com/example/docs/entity/model/MiniGolemEntityModel.java)
+
+The `MiniGolemEntityModel` class defines the visual model for a Mini Golem entity. It extends EntityModel, specifying how the entity's body parts (head, left leg, and right leg) are named.
```
```suggestion
The `MiniGolemEntityModel` class defines the visual model for a Mini Golem entity. It extends EntityModel, specifying how the entity's body parts (body, head, left leg, and right leg) are named.
```

Earthcomputer commented on develop/entities/first-entity.md:
```diff
@@ -0,0 +1,163 @@
+
+This method defines the Mini Golem's 3D model by creating its body, head, and legs as cuboids, setting their positions and texture mappings, and returning a `LayerDefinition` for rendering.
+
+Each part is added with a offset point for proper animation and alignment, ensuring the model appears correctly in-game.
```
This is far too vague. Crucially, the offset point is the origin for any transformations (such as rotation) applied to this model part. This is why you would want to ever change this offset point. All other coordinates in the model part are measured relative to this offset point.

JaaiDead replied:
Resolved

Earthcomputer commented on develop/entities/first-entity.md:
```diff
@@ -0,0 +1,163 @@
+### Setting up the Texture {#setting-up-texture}
+
+::: warning
+The size of the texture should match the values in the `LayerDefinition.create(modelData, 64, 32);`, 64 pixels wide and 32 pixels tall.
```
```suggestion
The size of the texture should match the values in the `LayerDefinition.create(modelData, 64, 32);`, 64 pixels wide and 32 pixels tall. If you need a bigger texture, then don't forget to change the size in `LayerDefinition.create` to match.
```

Earthcomputer commented on develop/entities/first-entity.md:
```diff
@@ -0,0 +1,163 @@
+
+@[code transcludeWith=:::model_animation](@/reference/latest/src/client/java/com/example/docs/entity/model/MiniGolemEntityModel.java)
+
+Most of the math is provided by Blockbench.
```
Huh? On the contrary, I believe I probably took this from one of vanilla's entities. It swings the legs based on a fairly simple transformation of a sine wave.

Does Blockbench commonly generate this? In which case I would change the wording to suggest that it *can* be generated by Blockbench, and instead describe the maths directly rather than a hand-wavy "Blockbench will do it for you" comment.

JaaiDead replied:
Resolved

Earthcomputer commented on develop/entities/first-entity.md:
```diff
@@ -0,0 +1,163 @@
+
+### Using Synched Entity Data {#using-synched-entity-data}
+
+We can store data on the entity itself by defining an `EntityDataAccessor` for it. This data will be synchronized with the client, making it a good place to define our dancing state.
```
This feels a little brief to me, although maybe I'm tripping. I think my problem with it is the way it's framed. As a mod developer of course you would do this as a subtask of making a dancing animation, but in a tutorial the important part is introducing the reader to synched entity data and explaining what it's all about. The same goes for the next section as well.

So rather than "we need a place to store our data, ah that's convenient there's this thing we can use to synchronize with the client", it should be "one way to synchronize entity data with the client is to use synched entity data. [Insert a brief introduction of synched entity data]. We will be using this for storing whether the entity is dancing, and synchronizing that information with the client. [Insert how we use synched entity data in this specific case".

Another problem with how this is worded is it implies that the primary purpose for synched entity data is to store data in an entity, when in fact it is to synchronize to the client. If anything doesn't need to be synchronized (and sometimes even if it does), you should use a field or another mechanism.

I think part of the problem is that this is included as a subsection of "adding a dancing animation", rather than a section in its own right. Even though it's part of setting up the animation, the dancing is part of the narrative of the whole tutorial, and only the actual animation itself needs to be under an animation header.

Earthcomputer commented on develop/entities/first-entity.md:
```diff
@@ -0,0 +1,163 @@
+
+We can store data on the entity itself by defining an `EntityDataAccessor` for it. This data will be synchronized with the client, making it a good place to define our dancing state.
+
+@[code transcludeWith=:::datatracker](@/reference/latest/src/main/java/com/example/docs/entity/MiniGolemEntity.java)
```
There is some animation state code interleaved in here. I think that should be left to the animation section later. (The field and the override of `onSyncedDataUpdated`.

Earthcomputer commented on develop/entities/first-entity.md:
```diff
@@ -0,0 +1,163 @@
+
+Looking into the game, you now have all you need to spawn the entity with `/summon example-mod:mini_golem`!
+
+![Spawn Egg showcase](/assets/develop/entity/mini_golem_summoned.png)
```
The alt text here is wrong

Earthcomputer commented on develop/entities/first-entity.md:
```diff
@@ -0,0 +1,163 @@
+
+Let's make this entity a bit more lively by giving it a dancing animation. We'll create a `MiniGolemAnimations` class with a simple animation that allows the Mini Golem to dance.
+
+@[code transcludeWith=:::dancing_animation](@/reference/latest/src/client/java/com/example/docs/entity/animation/MiniGolemAnimations.java)
```
As explained in the big long comment, I think this part should go in its own section at the end (just before the spawn egg).

cassiancc committed:
Apply suggestions from code review

Co-authored-by: Joseph Burton <burtonjae@hotmail.co.uk>

cassiancc committed:
Merge remote-tracking branch 'origin/main' into pr/267

cassiancc committed:
merge spawn egg article

cassiancc committed:
Update ModItems.java

its-miroma commented on .vitepress/sidebars/develop.ts:
```diff
@@ -128,6 +128,10 @@ export default [
     items: [
+      {
+        text: "develop.entities.first_entity",
+        link: "/develop/entities/first-entity", 
```
```suggestion
        link: "/develop/entities/first-entity",
```

its-miroma commented on reference/latest/src/client/java/com/example/docs/datagen/ExampleModEnglishLangProvider.java:
```diff
@@ -49,12 +50,16 @@ public void generateTranslations(HolderLookup.Provider wrapperLookup, Translatio
 		translationBuilder.add(ModBlocks.PRISMARINE_LAMP, "Prismarine Lamp");
 		translationBuilder.add(ModBlocks.ENGINE_BLOCK, "Engine Block");
 
+		translationBuilder.add(ModItems.MINI_GOLEM_SPAWN_EGG, "Mini Golem Spawn Egg");
```
This should be moved above, next to the other ModItems.
Also, please change the lines `translationBuilder.add("item.example-mod.*` to use `ModItems.*`.

JaaiDead replied:
Resolved

its-miroma reviewed (changes requested):
- Maybe the Spawn Egg Item page can now reference our custom Mini Golem, instead of frogs
- Entity Attributes' code blocks can now be merged I'd think?

cassiancc committed:
Merge branch 'main' into pr/267

cassiancc committed:
Port to 1.21.11

cassiancc committed:
spotless

cassiancc committed:
Apply new checkstyle rules

cassiancc committed:
Update develop.ts

JaaiDead committed:
Adds Graphing

JaaiDead committed:
Minor Fix

JaaiDead committed:
Reframeing Synched Data

JaaiDead committed:
Changeing Dancing Animation

Earthcomputer committed:
Merge branch 'main' into entities

JaaiDead committed:
Improves entity animation syncing and doc clarity

JaaiDead committed:
Removes unnecessary CSS whitespace for cleaner styles

JaaiDead committed:
Clarifies image caption for spawned entity

cassiancc commented on develop/entities/first-entity.md:
```diff
@@ -0,0 +1,221 @@
+- `Zombie` and `Skeleton` are all examples of a `Monster`, a variant of an `Entity` that is hostile to the `Player`.
+- `Minecart` and `Boat` are examples of a `VehicleEntity`, which has special logic for accepting player input.
+
+This tutorial will walk you through the creation of a `PathfinderEntity`, used by most mobs with pathfinding like `Zombie` and `Villager` alike.
```
```suggestion
This tutorial will walk you through the creation of a `PathfinderMob`, used by most mobs with pathfinding like `Zombie` and `Villager` alike.
```

Earthcomputer replied:
```suggestion
This tutorial will walk you through the creation of a `PathfinderEntity`, used by most mobs with pathfinding like `Zombie` and `Villager`.
```

cassiancc commented on develop/entities/first-entity.md:
```diff
@@ -0,0 +1,221 @@
+
+@[code transcludeWith=:::model_animation](@/reference/latest/src/client/java/com/example/docs/entity/model/MiniGolemEntityModel.java)
+
+At First , We apply the yaw and pitch to the head model part.
```
```suggestion
To start, apply the yaw and pitch to the head model part.
```

cassiancc commented on develop/entities/first-entity.md:
```diff
@@ -0,0 +1,221 @@
+
+To store data on an entity, the normal way is to simply add a field in entity class.
+
+Sometimes you need data from the server-side entity to be sycned with the client-side entity. See The [Networking Page](../networking) For More Info On The Client - Server Architecture. To do this we can define an `EntityDataAccessor` for it.
```
```suggestion
Sometimes you need data from the server-side entity to be synced with the client-side entity. See the [Networking Page](../networking) for more info on the Client-Server Architecture. To do this we can define an `EntityDataAccessor` for it.
```

Earthcomputer replied:
```suggestion
Sometimes you need data from the server-side entity to be synced with the client-side entity. See the [Networking Page](../networking) for more info on the client-server architecture. To do this we can use _synched data_ \[sic] by defining an `EntityDataAccessor` for it.
```

its-miroma commented on develop/entities/first-entity.md:
```diff
@@ -0,0 +1,221 @@
+- The `0.2` constant in the formula controls the frequency of the cosine wave (how fast the legs are swinging). Higher values result in a higher frequency.
+- The `1.4` constant in the formula controls the amplitude of the cosine wave (how far the legs swing). Higher values result in a higher amplitude.
+- The `limbSwingAmplitude` variable also affects the amplitude in the same way as the `1.4` constant. This variable changes based on the velocity of the entity, so that the legs swing more when the entity is moving faster and swing less or not at all when the entity is moving slower or not moving.
+- The `Mth.PI` constant for the left leg translates the cosine wave a half phase so that the left leg is swinging the opposite direction to the right leg.
```
```suggestion
- The `Math.PI` constant for the left leg translates the cosine wave a half phase so that the left leg is swinging the opposite direction to the right leg.
```

Earthcomputer replied:
Nope this is not a typo. It's `Mth.PI`.

cassiancc replied:
`Mth.PI` just calls `Math.PI`, is there a reason to use the Minecraft one?

Earthcomputer replied:
`Mth.PI` is `(float) Math.PI`, there's no need to downcast to float if you use `Mth.PI`.

its-miroma replied:
My fault, thought it was a typo

its-miroma commented on develop/entities/first-entity.md:
```diff
@@ -0,0 +1,221 @@
+
+![Graph](/assets/develop/entity/graphs/light_graph.png){.light-only}
+
+Feel free to play around with the constants to see how they affect the curve. [Desmos Link](https://www.desmos.com/calculator/bhlkblxhur)
```
```suggestion
Feel free to [play around with the constants on Desmos](https://www.desmos.com/calculator/bhlkblxhur) to see how they affect the curve.
```

is this an immutable "snapshot"?

Earthcomputer replied:
It's immutable, yes

Earthcomputer replied:
@JaaiDead this has not been fully resolved

JaaiDead replied:
ok 👍

its-miroma commented on develop/entities/first-entity.md:
```diff
@@ -0,0 +1,221 @@
+
+When the animation is playing we apply the animation, otherwise we use the old leg animation code.
+
+## Adding the Spawn Egg {#adding-spawn-egg}
```
Check out what we did in #518 — we don't have to repeat the code or list the steps I think.

its-miroma commented on develop/entities/first-entity.md:
```diff
@@ -0,0 +1,221 @@
+---
+title: Creating Your First Entity
+description: Learn how to register a simple entity and how to give it goals, render, model, and animate it.
+authors:
```
This article does not come from the wiki at all, right?

Earthcomputer replied:
No, it's brand new

its-miroma reviewed (changes requested):
this is a quick glance-over, not a full review. Also, I think my previous review's subject wasn't fully acknowledged yet?

> - ~~Maybe the Spawn Egg Item page can now reference our custom Mini Golem, instead of frogs~~
> - Entity Attributes' code blocks can now be merged I'd think?

Earthcomputer commented on develop/entities/first-entity.md:
```diff
@@ -0,0 +1,221 @@
+An entity is a movable object in a world with logic attached to them. A few examples include:
+
+- `Villager`, `Pig`, and `Goat` are all examples of a `Mob`, the most common type of entity - something alive.
+- `Zombie` and `Skeleton` are all examples of a `Monster`, a variant of an `Entity` that is hostile to the `Player`.
```
```suggestion
- `Zombie` and `Skeleton` are examples of a `Monster`, a variant of an `Entity` that is hostile to the `Player`.
```

Earthcomputer commented on develop/entities/first-entity.md:
```diff
@@ -0,0 +1,221 @@
+
+The first step in creating a custom entity is defining its class and registering it with the game.
+
+We'll create the class `MiniGolemEntity` for our entity, and start by giving it attributes. [Attributes](attributes) decide the Maximum Health, Movement Speed & Tempt Range of the entity.
```
```suggestion
We'll create the class `MiniGolemEntity` for our entity, and start by giving it attributes. [Attributes](attributes) decide various things including the maximum health, movement speed, and tempt range of the entity.
```

Earthcomputer commented on develop/entities/first-entity.md:
```diff
@@ -0,0 +1,221 @@
+
+@[code transcludeWith=:::registerclass](@/reference/latest/src/main/java/com/example/docs/entity/MiniGolemEntity.java)
+
+To register your entity, it's recommended to create a separate class, `ModEntityTypes`, where you register any and all entity types, set their dimensions, and register their attributes.
```
```suggestion
To register your entity, it's recommended to create a separate class, `ModEntityTypes`, where you register any and all entity types, set their sizes, and register their attributes.
```

Earthcomputer commented on develop/entities/first-entity.md:
```diff
@@ -0,0 +1,221 @@
+
+## Adding Goals {#adding-goals}
+
+Goals are the system that handle an entity's objective/aim, providing them with a defined set of behaviour. Goals have a certain priority: goals with a lower value for the priority are prioritized over goals with a higher value for the priority.
```
```suggestion
Goals are the system that handle an entity's objective/aim, providing them with a defined set of behavior. Goals have a certain priority: goals with a lower value for the priority are prioritized over goals with a higher value for the priority.
```

Earthcomputer commented on develop/entities/first-entity.md:
```diff
@@ -0,0 +1,221 @@
+::: info
+
+1. `TemptGoal` - The entity is attracted towards a player holding an item.
+2. `RandomStrollGoal` - Walks/Wanders around the world.
```
```suggestion
2. `RandomStrollGoal` - Walks/wanders around the world.
```

Earthcomputer commented on develop/entities/first-entity.md:
```diff
@@ -0,0 +1,221 @@
+
+1. `TemptGoal` - The entity is attracted towards a player holding an item.
+2. `RandomStrollGoal` - Walks/Wanders around the world.
+3. `LookAtPlayerGoal` - To look at the `Cow` entity.
```
?

Earthcomputer commented on develop/entities/first-entity.md:
```diff
@@ -0,0 +1,221 @@
+
+## Creating Rendering {#creating-rendering}
+
+Rendering refers to the process of converting game data such as blocks, entities, and environments into visual representations displayed on the player's screen. This involves determining how objects are illuminated, shaded, and textured.
```
While true, illumination and shading (in the natural sense of the word, not "shaders") aren't directly related to entity rendering in most cases.

Earthcomputer commented on develop/entities/first-entity.md:
```diff
@@ -0,0 +1,221 @@
+
+@[code transcludeWith=:::entitystate](@/reference/latest/src/client/java/com/example/docs/entity/state/MiniGolemEntityRenderState.java)
+
+The render state determines how the entity is visually represented, including animation states such as movement and idle behaviors.
```
The render state actually just stores the state required for rendering, as the name suggests. It does not determine *how* the entity is rendered.

Earthcomputer commented on develop/entities/first-entity.md:
```diff
@@ -0,0 +1,221 @@
+
+@[code transcludeWith=:::model1](@/reference/latest/src/client/java/com/example/docs/entity/model/MiniGolemEntityModel.java)
+
+The `MiniGolemEntityModel` class defines the visual model for a Mini Golem entity. It extends EntityModel, specifying how the entity's body parts (body, head, left leg, and right leg) are named.
```
```suggestion
The `MiniGolemEntityModel` class defines the visual model for a Mini Golem entity. It extends `EntityModel`, specifying how the entity's body parts (body, head, left leg, and right leg) are named.
```

Earthcomputer commented on develop/entities/first-entity.md:
```diff
@@ -0,0 +1,221 @@
+
+This method defines the Mini Golem's 3D model by creating its body, head, and legs as cuboids, setting their positions and texture mappings, and returning a `LayerDefinition` for rendering.
+
+Each part is added with an offset point which is the origin for all of the transformations applied to that part. All other coordinates in the model part are measured relative to this offset point.Most of the math is provided by Blockbench.
```
```suggestion
Each part is added with an offset point which is the origin for all of the transformations applied to that part. All other coordinates in the model part are measured relative to this offset point.
```

Earthcomputer commented on develop/entities/first-entity.md:
```diff
@@ -0,0 +1,221 @@
+
+:::
+
+We'll now need to create a `ModEntityModelLayers` class in the client package. This entity only has a single texture layer, but other entities may use multiple - think of the secondary skin layer on mobs like the `Player` or a `Spider`'s eyes.
```
```suggestion
We'll now need to create a `ModEntityModelLayers` class in the client package. This entity only has a single texture layer, but other entities may use multiple - think of the secondary skin layer on entities like the `Player` or a `Spider`'s eyes.
```

Earthcomputer commented on develop/entities/first-entity.md:
```diff
@@ -0,0 +1,221 @@
+
+::: warning
+
+The size of the texture should match the values in the `LayerDefinition.create(modelData, 64, 32);`, 64 pixels wide and 32 pixels tall. If you need a bigger texture, then don't forget to change the size in `LayerDefinition.create` to match.
```
```suggestion
The size of the texture should match the values in the `LayerDefinition.create(modelData, 64, 32);`, 64 pixels wide and 32 pixels tall. If you need a differently-sized texture, then don't forget to change the size in `LayerDefinition.create` to match.
```

Earthcomputer commented on develop/entities/first-entity.md:
```diff
@@ -0,0 +1,221 @@
+
+:::
+
+Each model part / box is expecting a net on the texture in a particular location. By default, it's expecting it at `0, 0` (the top left), but this can be changed by calling the texOffs function in CubeListBuilder.
```
```suggestion
Each model part / box is expecting a net on the texture in a particular location. By default, it's expecting it at `0, 0` (the top left), but this can be changed by calling the `texOffs` function in `CubeListBuilder`.
```

Earthcomputer commented on develop/entities/first-entity.md:
```diff
@@ -0,0 +1,221 @@
+
+Each model part / box is expecting a net on the texture in a particular location. By default, it's expecting it at `0, 0` (the top left), but this can be changed by calling the texOffs function in CubeListBuilder.
+
+For example purposes, you can use this example texture for `assets/example-mod/textures/entity/mini_golem.png`
```
```suggestion
For example purposes, you can use this texture for `assets/example-mod/textures/entity/mini_golem.png`
```

Earthcomputer commented on develop/entities/first-entity.md:
```diff
@@ -0,0 +1,221 @@
+
+### Creating the Renderer {#creating-the-renderer}
+
+A mob's entity renderer enables you to view your entity in-game. We'll create a new class, `MiniGolemEntityRenderer`, which will tell Minecraft what texture, model and entity render state to use for this entity.
```
```suggestion
An entity's renderer enables you to view your entity in-game. We'll create a new class, `MiniGolemEntityRenderer`, which will tell Minecraft what texture, model and entity render state to use for this entity.
```

Earthcomputer commented on develop/entities/first-entity.md:
```diff
@@ -0,0 +1,221 @@
+
+Sometimes you need data from the server-side entity to be sycned with the client-side entity. See The [Networking Page](../networking) For More Info On The Client - Server Architecture. To do this we can define an `EntityDataAccessor` for it.
+
+For our entity we'll make it dance ever so often, so we need to create a dancing state that is synchronized with the client so that it can be animated later. However, the dancing cooldown need not be syched with the client.
```
```suggestion
For our entity we'll make it dance every so often, so we need to create a dancing state that is synchronized with the client so that it can be animated later. However, the dancing cooldown need not be synced with the client.
```

Earthcomputer commented on develop/entities/first-entity.md:
```diff
@@ -0,0 +1,221 @@
+
+## Storing Data to NBT {#storing-data-to-nbt}
+
+For persistent data that can be saved after the game is closed, we'll start by overriding the `addAdditionalSaveData` and `readAdditionalSaveData` methods in `MiniGolemEntity`. We can use this to store the amount of time remaining in the dancing animation.
```
```suggestion
For persistent data that can be saved after the game is closed, we will override the `addAdditionalSaveData` and `readAdditionalSaveData` methods in `MiniGolemEntity`. We can use this to store the amount of time remaining in the dancing animation.
```

Earthcomputer commented on develop/entities/first-entity.md:
```diff
@@ -0,0 +1,221 @@
+
+@[code transcludeWith=:::dancing_animation](@/reference/latest/src/main/java/com/example/docs/entity/MiniGolemEntity.java)
+
+We have overridden the `onSyncedDataUpdated` method. This gets called whenever synched data is updated both the server and the client. The if-statement check wether the synched data that was updated is the dancing synched data.
```
```suggestion
We have overridden the `onSyncedDataUpdated` method. This gets called whenever synched data is updated both the server and the client. The if-statement checks whether the synched data that was updated is the dancing synched data.
```

cassiancc committed:
Apply suggestions from code review

Co-authored-by: Joseph Burton <burtonjae@hotmail.co.uk>

JaaiDead committed:
Improves clarity of Desmos link instruction.

cassiancc committed:
Apply suggestions

Co-Authored-By: Joseph Burton <burtonjae@hotmail.co.uk>

JaaiDead committed:
Updates spawn egg to Mini Golem entity.

Replaces references to a generic spawn egg with Mini Golem-specific details in documentation, source code, and assets.

JaaiDead committed:
Linting Fix

JaaiDead committed:
Merge branch 'main' into entities

JaaiDead committed:
Remove unnecessary blank lines in ExampleModModelProvider and ModItems

JaaiDead committed:
Deleted Datagen files.

cassiancc committed:
Minor revisions

JaaiDead committed:
Merge branch 'main' into entities

JaaiDead commented:
why GitHub dismissed the review huh, was resolving conflicts

cassiancc commented:
Any change to the PR dismisses any approving reviews.

Earthcomputer commented:
@JaaiDead there are failing checks

JaaiDead committed:
Build

JaaiDead committed:
Undo Commit.

JaaiDead reviewed (commented):
Good to Go :D.

its-miroma commented on reference/latest/src/client/java/com/example/docs/datagen/ExampleModModelProvider.java:
```diff
@@ -96,14 +96,13 @@ public void generateItemModels(ItemModelGenerators itemModelGenerator) {
 		//:::generated
 
+		// :::datagen-model:provider
+
```
```suggestion
```

its-miroma commented on reference/latest/src/client/java/com/example/docs/entity/model/MiniGolemEntityModel.java:
```diff
@@ -0,0 +1,99 @@
+		PartDefinition root = modelData.getRoot();
+		root.addOrReplaceChild(
+				PartNames.BODY,
+				CubeListBuilder.create().addBox(-6, -6, -6, 12, 12, 12), // x, y, z is the dimensions | width, height, depth
```
Just as an idea, do you think this is clearer? feel free to ignore

```suggestion
				CubeListBuilder.create().addBox(
						/* x */ -6,
						/* y */ -6,
						/* z */ -6,
						/* width */ 12,
						/* height */ 12,
						/* depth */ 12),
```

its-miroma commented on develop/entities/first-entity.md:
```diff
@@ -0,0 +1,205 @@
+- `Zombie` and `Skeleton` are examples of a `Monster`, a variant of an `Entity` that is hostile to the `Player`.
+- `Minecart` and `Boat` are examples of a `VehicleEntity`, which has special logic for accepting player input.
+
+This tutorial will walk you through the creation of a `PathfinderMob`, used by most mobs with pathfinding like `Zombie` and `Villager`.
```
```suggestion
Entities are dynamic, interactive objects in the game that are not part of the terrain (like blocks). Entities can move around and interact with the world in various ways. A few examples include:

- `Villager`, `Pig`, and `Goat` are all examples of a `Mob`, the most common type of entity - something alive.
- `Zombie` and `Skeleton` are examples of a `Monster`, a variant of an `Entity` that is hostile to the `Player`.
- `Minecart` and `Boat` are examples of a `VehicleEntity`, which has special logic for accepting player input.

This tutorial will walk you through the process of creating a custom _Mini Golem_. This entity will have fun animations. It will be a `PathfinderMob`, which is the class used by most mobs with pathfinding, such as `Zombie` and `Villager`.
```

its-miroma commented on develop/entities/first-entity.md:
```diff
@@ -0,0 +1,205 @@
+
+Rendering refers to the process of converting game data such as blocks, entities, and environments into visual representations displayed on the player's screen. This involves determining how objects are illuminated, shaded, and textured.
+
+::: tip
```
```suggestion
::: info
```

its-miroma commented on develop/entities/first-entity.md:
```diff
@@ -0,0 +1,205 @@
+
+The higher Y values in the model, the lower you are in the entity. This is the reverse compared to in-game coordinates.
+
+:::
```
```suggestion
::: warning

Higher Y values in the model correspond to the **bottom** of the entity. This is the reverse compared to in-game coordinates.

:::
```

its-miroma commented on develop/entities/first-entity.md:
```diff
@@ -0,0 +1,205 @@
+
+The size of the texture should match the values in the `LayerDefinition.create(modelData, 64, 32);`, 64 pixels wide and 32 pixels tall. If you need a differently-sized texture, then don't forget to change the size in `LayerDefinition.create` to match.
+
+:::
```
```suggestion
::: tip

The size of the texture should match the values in the `LayerDefinition.create(modelData, 64, 32)`: 64 pixels wide and 32 pixels tall. If you need a differently-sized texture, then don't forget to change the size in `LayerDefinition.create` to match.

:::
```

its-miroma commented on develop/entities/first-entity.md:
```diff
@@ -0,0 +1,205 @@
+
+![Graph](/assets/develop/entity/graphs/light_graph.png){.light-only}
+
+Feel free to [play around with the constants on Desmos](https://www.desmos.com/calculator/bhlkblxhur) to see how they affect the curve.
```
```suggestion
You can plot those on a graph to see what they look like:

![Graph](/assets/develop/entity/graphs/dark_graph.png){.dark-only}
![Graph](/assets/develop/entity/graphs/light_graph.png){.light-only}

The "red" curve is for the left leg, and the "black" one is for the right. The horizontal x-axis represents time, and the y-axis indicates the angle of the leg limbs.

Feel free to [play around with the constants on Desmos](https://www.desmos.com/calculator/bhlkblxhur) to see how they affect the curve.
```

waiting on the changes requested here:
https://discordapp.com/channels/507304429255393322/1208846408552030238/1484908219158499430 via https://discord.fabricmc.net/

JaaiDead replied:
Done

its-miroma commented on develop/entities/first-entity.md:
```diff
@@ -0,0 +1,205 @@
+
+Sometimes you need data from the server-side entity to be synced with the client-side entity. See the [Networking Page](../networking) for more info on the client-server architecture. To do this we can use _synched data_ \[sic] by defining an `EntityDataAccessor` for it.
+
+For our entity we'll make it dance every so often, so we need to create a dancing state that is synchronized with the client so that it can be animated later. However, the dancing cooldown need not be synced with the client.
```
```suggestion
In our case we want our entity to dance every so often, so we need to create a dancing state that is synchronized between the clients so that it can be animated later. However, the dancing cooldown need not be synced with the client because the animation is triggered by the server.
```

correct me if I'm wrong

its-miroma commented on develop/entities/first-entity.md:
```diff
@@ -0,0 +1,205 @@
+
+## Adding the Spawn Egg {#adding-spawn-egg}
+
+To add the new spawn egg of the Mini Golem entity, refer to the full article on [Creating a Spawn Egg](../items/spawn-egg).
```
```suggestion
To add a spawn egg for the Mini Golem entity, refer to the full article on [Creating a Spawn Egg](../items/spawn-egg).
```

its-miroma commented on develop/entities/first-entity.md:
```diff
@@ -0,0 +1,205 @@
+  - We have used linear interpolation, which is the simplest and changes the value (in our case rotation of the model part) at a constant rate from one keyframe to the next. Vanilla also provides Catmull-Rom spline interpolation, which produces a smoother transition between keyframes.
+  - Modders can also create custom interpolation types.
+
+Finally, let's hook up the animation to the model.
```
```suggestion
Finally, let's hook up the animation to the model:
```

JaaiDead committed:
Apply suggestions from code review

Co-authored-by: Miroma <its.miroma@proton.me>

JaaiDead committed:
Apply suggestions from code review

Co-authored-by: Miroma <its.miroma@proton.me>

JaaiDead committed:
Code changes.

its-miroma commented on develop/entities/first-entity.md:
```diff
@@ -0,0 +1,200 @@
+
+The "red" curve is for the left leg, and the "black" one is for the right. The horizontal x-axis represents time, and the y-axis indicates the angle of the leg limbs.
+
+Feel free to [play around with the constants on Desmos](https://www.desmos.com/calculator/bhlkblxhur) to see how they affect the curve.
```
Let's update the graph link and the screenshots, then LGTM!

JaaiDead committed:
update graph links

JaaiDead committed:
update Desmos link and replace graph images.

JaaiDead committed:
Updates Desmos link.

JaaiDead reviewed (commented):
Approved
